File: src/precision.rs
use std::fmt;

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Default, Serialize, Deserialize)]
pub enum Precision {
    Float16,
    #[default]
    BFloat16,
    Float32,
}

impl Precision {
    pub const ALL: &'static [Self] = &[Self::Float16, Self::BFloat16, Self::Float32];
}

impl fmt::Display for Precision {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(match self {
            Precision::Float16 => "float16",
            Precision::BFloat16 => "bfloat16",
            Precision::Float32 => "float32",
        })
    }
}

File: server/image/precision.py
import torch

from enum import Enum


class Precision(Enum):
    FLOAT16 = 0
    BFLOAT16 = 1
    FLOAT32 = 2

    def parse(precision: str) -> "Precision":
        return {
            "float16": Precision.FLOAT16,
            "bfloat16": Precision.BFLOAT16,
            "float32": Precision.FLOAT32,
        }[precision]

    def dtype(self):
        match self:
            case Precision.FLOAT16:
                return torch.float16
            case Precision.BFLOAT16:
                return torch.bfloat16
            case Precision.FLOAT32:
                return torch.float32
